### PR TITLE
Remove list argument broadcasting and simplify transpile()

### LINKED ISSUE #7741: Big overhead in parallel transpile for large target systems
### Environment

- **Qiskit Terra version**: `0.19.2`
- **Python version**: 3.8
- **Operating system**: Linux

### What is happening?

There is a large overhead when doing parallel transpilation for circuits targeting larger systems.  In the case below they are targeting the 127Q system, but the circuits themselves consist of only `x` gates and are already the correct width.  Thus transpilation should just be a pass-through.  However, in parallel mode (quad-core machine) it takes 330sec to transpile.  If I turn off parallel transpile it takes 3sec.

```python

from qiskit import *
from mthree.circuits import balanced_cal_strings, balanced_cal_circuits

b_strs = balanced_cal_strings(127)
cal_circs = balanced_cal_circuits(b_strs)

backend = provider.backend.ibm_washington
trans_circs = transpile(cal_circs, backend, optimization_level=0)
```

### How can we reproduce the issue?

Run above

### What should happen?

Parallel should not give 100x overhead. 

### Any suggestions?

I am guessing this is copying overhead from forking when the systems are large.

### LINKED ISSUE #7883: Seed transpiler takes list but it is not in the docstrings
### Environment

- **Qiskit Terra version**: 0.20
- **Python version**: 
- **Operating system**: 

### What is happening?

In order to seed transpiler for a list of circuits a list of seeds is required.  This is supported, but not mentioned anywhere:

https://github.com/Qiskit/qiskit-terra/blob/a296ca009bf440b5a2cb00f61b8221c6ce9aa044/qiskit/compiler/transpiler.py#L65

### How can we reproduce the issue?

See the docstring

### What should happen?

It should be listed as an option because I originally assumed having all processes set with a single int seed was a bug.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit updates the transpile() function to no longer support broadcast of lists of arguments. This functionality was deprecated in the 0.23.0 release. As part of this removal the internals of the transpile() function are simplified so we don't need to handle broadcasting, building preset pass managers, parallel dispatch, etc anymore as this functionality (without broadcasting) already exists through the transpiler API. Besides greatly simplifying the transpile() code and using more aspects of the public APIs that exist in the qiskit.transpiler module, this commit also should fix the overhead we have around parallel execution due to the complexity of supporting broadcasting. This overhead was partially addressed before in #7789 which leveraged shared memory to minimize the serialization time necessary for IPC but by using `PassManager.run()` internally now all of that overhead is removed as the initial fork will have all the necessary context in each process from the start.

Three seemingly unrelated changes made here were necessary to support our current transpile() API without relying on custom pass manager construction.

~The first is the handling of layout from intlist. The current Layout class is dependent on a circuit because it maps Qubit~
~objects to a physical qubit index. Ideally the layout structure would just map virtual indices to physical indices (see #8060~
~for a similar issue, also it's worth noting this is how the internal NLayout and QPY represent layout), but because of the~
~existing API the construction of a Layout is dependent on a circuit. For the initial_layout argument when running with~
~multiple circuits to avoid the need to broadcasting the layout construction for supported input types that need the circuit~
~to lookup the Qubit objects the SetLayout pass now supports taking in an int list and will construct a Layout object at run~
~time. This effectively defers the Layout object creation for initial_layout to run time so it can be built as a function of the~
~circuit as the API demands.~ (this was handled separately in https://github.com/Qiskit/qiskit-terra/pull/10344)

The second is the FakeBackend class used in some tests was constructing invalid backends in some cases. This wasn't caught in the previous structure because the backends were not actually being parsed by transpile() previously which masked this issue. This commit fixes that issue because PassManagerConfig.from_backend() was failing because of the invalid backend construction.

The third issue is a new _skip_target private argument to generate_preset_pass_manager() and PassManagerConfig. This was necessary to recreate the behavior of transpile() when a user provides a BackendV2 and either `basis_gates` or `coupling_map` arguments. In general the internals of the transpiler treat a target as higher priority because it has more complete and restrictive constraints than the basis_gates/coupling map objects. However, for transpile() if a backendv2 is passed in for backend paired with coupling_map and/or basis_gates the expected workflow is that the basis_gates and coupling_map arguments take priority and override the equivalent attributes from the backend. To facilitate this we need to block pulling the target from the backend This should only be needed for a short period of time as when #9256 is implemented we'll just build a single target from the arguments as needed.

### Details and comments

Fixes #7741
Fix #7883 

TODO:

- [x] Fix last failing tests (dt handling for scheduling)

### CODE REVIEW

File: qiskit/providers/fake_provider/fake_backend.py
Comment (jlapeyre) [inline qiskit/providers/fake_provider/fake_backend.py:481]:
```suggestion
        if coupling_map is None:
            return None
        unique_qubits = list(set().union(*coupling_map))
```

**Review by danielleodigie — COMMENTED** [review]
From what I can see, it LGTM, other than that failing test I guess. I think it's a Linux test, and I can't recreate the issue on my end.

**Comment by jakelishman** [comment]
The test failure was a flakiness in the CI suite that I've made #10439 to address.

File: releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml
Comment (jakelishman) [inline releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml:10]: We maybe want to make a little clearer that passing in a list of circuit inputs is still permitted (for the parallelisation).

File: test/python/compiler/test_transpiler.py
Comment (jakelishman) [inline test/python/compiler/test_transpiler.py:639]: Is the new error message significantly worse?

File: releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml
Comment (jakelishman) [inline releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml:40]: Super minor, but this seems like an unusual pattern to suggest compared to `results = [transpile(qc, coupling_map=cm) for cm in cmap]` (and corresponding removal of the `itertools` import).

File: qiskit/compiler/transpiler.py
Comment (jakelishman) [inline qiskit/compiler/transpiler.py:323]: Minor: this probably wants to maintain the `is None` checks, because `basis_gates=[]` is an explicit (pathological) setting of `basis_gates` that should probably be respected.

File: qiskit/compiler/transpiler.py
Comment (jakelishman) [inline qiskit/compiler/transpiler.py:329]: `_parse_approximation_degree` below still appears to have list-based handling, but I think it's meant to be scalar only.

File: qiskit/compiler/transpiler.py
Comment (jakelishman) [inline qiskit/compiler/transpiler.py:331]: `output_name` still broadcasts in `_parse_output_name`, but that at least doesn't get passed to the `PassManager`, so I guess we're ok with keeping that?  I honestly had no idea we even had that option.

File: qiskit/compiler/transpiler.py
Comment (jakelishman) [inline qiskit/compiler/transpiler.py:335]: Feels like `_check_circuits_coupling_map` doesn't really need `cmap_conf` to be a list with `n` copies of the same Python object, and `cmap_conf` isn't used elsewhere.  Could we just pass the single `CouplingMap`?

File: qiskit/compiler/transpiler.py
Comment (jakelishman) [inline qiskit/compiler/transpiler.py:353]: To me, this seems to be indicating a problem with the passes that interpret this data; they should read it from the circuit during their `run` methods.  It's the same problem that we had with `SetLayout` until just recently.

Ideally we'd fix that before this PR, but there's no time before 0.25 now.  Can we check with the people who might be using these options within IBM that they're not going to be hamstrung by the decay to serial transpile, at least?

File: qiskit/compiler/transpiler.py
Comment (jakelishman) [inline qiskit/compiler/transpiler.py:918]: Noooo, my first non-trivial merged PR to Terra D: (#6814, haha)

**Review by jakelishman — COMMENTED** [review]
This looks a jillion times better than the old code.  Given that this PR claims to close #7741, do you have approximate timings for the type of thing in that issue?

File: qiskit/compiler/transpiler.py
Comment (mtreinish) [reply qiskit/compiler/transpiler.py:353]: It is definitely a problem. My concern with doing the same fix as with `SetLayout` was I didn't think it was used in a single place like the input layout and that `InstructionDurations` would have to be updated in an instance variable in multiple passes. So I went with this approach instead to avoid doing this in multiple places.

@nkanazawa1989 or @itoko do you have any idea/thoughts  here regarding the potential impact of serializing the execution with a custom instruction durations or dt input? I expect you've got the best handle on the impact of setting custom instruction durations (or dt) on the argument inputs to `transpile()`.

My thinking is for most users it won't be noticeable because my feeling is most of them are on macOS anyway and we disable multiprocessing by default on macOS.

**Comment by mtreinish** [comment]
> This looks a jillion times better than the old code. Given that this PR claims to close #7741, do you have approximate timings for the type of thing in that issue?

I don't have the timings handy. I think that I tried to run the recreate from the issue when I wrote this a month ago and I was having trouble reproducing the issue locally. I'll give it a try again in a bit and see if I can get real numbers.

File: test/python/compiler/test_transpiler.py
Comment (mtreinish) [reply test/python/compiler/test_transpiler.py:639]: It was when I modified this test, but @1ucian0 added a real error message in `SetLayout` as part of #10344 so it'll say:
```
The length of the layout is different than the size of the circuit 1024 <> 2
```
now. (I glossed over `<>` in my review we should change that to actually say `1024 qubits in the layout != 2 qubits in the circuit` or something more explicit instead)

File: releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml
Comment (mtreinish) [reply releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml:40]: TBH, I don't know why I did this, I think I just copied it blindly from the deprecation release note (although I wrote that too). But yeah it makes sense to update this

File: qiskit/compiler/transpiler.py
Comment (mtreinish) [reply qiskit/compiler/transpiler.py:329]: Yeah good catch, we don't need to check the list anymore as that's invalid input now.

File: qiskit/compiler/transpiler.py
Comment (mtreinish) [reply qiskit/compiler/transpiler.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/10291/commits/75c918b3beaa851c2f121049c2df5365c76f7203

File: test/python/compiler/test_transpiler.py
Comment (jakelishman) [reply test/python/compiler/test_transpiler.py:639]: or just stick `from __future__ import barry_as_FLUFL` at the top of the file lol

File: qiskit/compiler/transpiler.py
Comment (mtreinish) [reply qiskit/compiler/transpiler.py:329]: Removed in: https://github.com/Qiskit/qiskit-terra/pull/10291/commits/75c918b3beaa851c2f121049c2df5365c76f7203

File: qiskit/compiler/transpiler.py
Comment (mtreinish) [reply qiskit/compiler/transpiler.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/10291/commits/0ea2edae911f6b9f02bc2dc627a0896853f4efc1 although my guess is the functional difference between `basis_gates = []` and `basis_gates = None` is the empty list will error 100% of the time in basis translation.

File: releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml
Comment (mtreinish) [reply releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml]: Updated in: https://github.com/Qiskit/qiskit-terra/pull/10291/commits/ffd10191defea7300c659fa1c4e3aa3187be3a26

File: releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml
Comment (mtreinish) [reply releasenotes/notes/remove-transpile-broadcast-1dfde28d508efa0d.yaml]: Updated in: https://github.com/Qiskit/qiskit-terra/pull/10291/commits/ffd10191defea7300c659fa1c4e3aa3187be3a26

File: qiskit/compiler/transpiler.py
Comment (mtreinish) [reply qiskit/compiler/transpiler.py:331]: Yeah it shouldn't matter for this, it felt like if we're taking an output name argument it needs to support broadcasting. Since if you did `transpile([qc1, qc2, qc3, qc4], output_name='foo')` I'm not sure what the expected behavior for that would be (or why we'd force it on people). Since the broadcasting doesn't effect the `PassManager` object or the parallel dispatch I figured it was fine to keep it.

Comment (jakelishman) [reply qiskit/compiler/transpiler.py:331]: I'm fine with keeping it broadcasting, since it doesn't affect the `PassManager`. If the option wasn't already in `transpile`, I wouldn't have been in favour of adding it, haha - turns out that the initial use-case was to distinguish the same circuit transpiled with broadcast arguments (#2745).

**Review by jakelishman — APPROVED** [review]
I'm happy with this and ready to merge, pending an answer from one of the potentially affected people about https://github.com/Qiskit/qiskit-terra/pull/10291#discussion_r1265917229.  In general, I think that's probably a trade-off we're going to need to make; the parallel overhead for pulse jobs was already (I think?) high enough that parallel transpile probably wasn't helping a great amount already, and the behaviour is still achievable with explicit parallelisation with built-in Python tools.

I'll hold off on tagging for merge until there's a resolution in that comment thread.

File: qiskit/compiler/transpiler.py
Comment (itoko) [reply qiskit/compiler/transpiler.py:352]: I think the potential impact of serializing the execution with a custom instruction durations or dt input would be little (at least for noisy simulation users). I didn't know until now that instruction durations were being used for pulse gates. I've been thinking `InstructionDurations` is for custom gates without calibrations used for noisy simulation and pulse gates are separately handled [here](https://github.com/Qiskit/qiskit-terra/blame/a804df58c8e59c101fc3c13ef6a1a4cb172b9a5b/qiskit/transpiler/passes/scheduling/scheduling/base_scheduler.py#L61). @naoki, do you think the case you mentioned above is covered by the code in the link without handling of `circuits` in `_parse_instruction_durations`?

**Pull request merged**
